commit dc37bb94858cf303642eaade6f761f5972faad00 (refs/changes/68/5168/4)
Author: Mike Gerdts <mike.gerdts@joyent.com>
Date:   2018-12-20T16:47:01+00:00 (10 months ago)
    
    OS-7414 runtest should alert if new cores are found
    Reviewed by: Josh Wilsdon <josh@wilsdon.ca>
    Approved by: Josh Wilsdon <josh@wilsdon.ca>

diff --git a/src/vm/runtest b/src/vm/runtest
index 086f568c..326fc38c 100755
--- a/src/vm/runtest
+++ b/src/vm/runtest
@@ -174,6 +174,9 @@ if ! imgadm info "$uefi_img_uuid" >/dev/null 2>&1; then
 	zfs destroy -r "$uefi_zvol"
 fi
 
+# Keep track of when we started so it is easy to identify new core files
+touch /tmp/runtest.start.$$
+
 set +o errexit
 set -o pipefail
 /usr/bin/ctrun -l child -o noorphan /usr/node/bin/node \
@@ -202,6 +205,11 @@ if [[ ${TEST_EXIT_CODE} != 0 ]]; then
     tests="?"
 fi
 
+coredir=$(cd $(dirname $(coreadm $$ | awk '{gsub("%Z", "/", $2); print $2;}')) \
+    && pwd -P)
+cores=$(find "$coredir" -type f -newer /tmp/runtest.start.$$)
+rm /tmp/runtest.start.$$
+
 if [[ -t 0 ]]; then
     # We're on a terminal so output the summary
     echo "#"
@@ -216,10 +224,18 @@ if [[ -t 0 ]]; then
     if [[ ${fail} -gt 0 ]]; then
         echo -e "# \033[31mFAIL: ${fail} / ${tests}\033[39m"
     fi
+    if [[ -n "$cores" ]]; then
+        echo -e "# \033[31mNew core files\033[39m"
+        # Send a list of core files to stdout for testers and `runtests` to see
+        # and another in this run's log.
+        for core in $cores; do
+            echo "# new core: $core"
+        done | tee -a /tmp/test.output.$$
+    fi
     echo "#"
 fi
 
-if [[ ${fail} -gt 0 ]]; then
+if [[ ${fail} -gt 0 || -n "$cores" ]]; then
     exit 1
 fi
 
diff --git a/src/vm/runtests b/src/vm/runtests
index 3ff6904b..47615488 100755
--- a/src/vm/runtests
+++ b/src/vm/runtests
@@ -55,6 +55,7 @@ elapsed=$((${end_time} - ${start_time}))
 tests=$(grep "^# tests [0-9]" /tmp/test.output.$$ | cut -d ' ' -f3 | xargs | tr ' ' '+' | bc)
 passed=$(grep "^# pass  [0-9]" /tmp/test.output.$$ | tr -s ' ' | cut -d ' ' -f3 | xargs | tr ' ' '+' | bc)
 skipped=$(grep "^# skip  [0-9]" /tmp/test.output.$$ | tr -s ' ' | cut -d ' ' -f3 | xargs | tr ' ' '+' | bc)
+cores=$(grep "^# new core:" /tmp/test.output.$$ | awk '{print $NF}')
 
 [[ -z ${tests} ]] && tests=0
 [[ -z ${passed} ]] && passed=0
@@ -91,6 +92,13 @@ if [[ -n ${failed_tests} ]]; then
         echo -e "#  ${FAIL_COLOR}${test}${NORMAL_COLOR}"
     done
 fi
+if [[ -n "$cores" ]]; then
+    echo "#"
+    echo "# ** NEW CORE FILES **"
+    for core in $cores; do
+        echo -e "#  ${FAIL_COLOR}${core}${NORMAL_COLOR}"
+    done
+fi
 echo "#"
 
 if [[ ${fail} -gt 0 || ${test_errors} -gt 0 ]]; then
